aidial_adapter_openai/app.py

```diff
@@ -100,15 +67,37 @@ def get_api_version(request: Request):
 
 @app.post("/openai/deployments/{deployment_id}/chat/completions")
 async def chat_completion(deployment_id: str, request: Request):
+
     data = await parse_body(request)
+    is_stream = bool(data.get("stream", False))
+    api_version = get_api_version(request)
 
-    is_stream = data.get("stream", False)
+    upstream_endpoint = request.headers["X-UPSTREAM-ENDPOINT"]
 
-    api_type, api_key = await get_credentials(request, chat_completions_parser)
+    # If it is not chat completions, but legacy completions
+    if not chat_completions_parser.is_valid(
```

Comment:
It's better to parse and validate at the same time, so that you don't need to keep in mind that every `parse` should be only called under the branch where `is_valid` is True:

```python
if (endpoint := completions_parser.parse(upstream_endpoint)) is not None:
    ...
```

Reply:
This would not work with current parsers, since they are raising `DialException` if parsing has not succeeded
I don't really want to change behaviour of this class 

---

Also, another problem, that raises when we send both `/chat/completions` and `/completions` to parser, is that with current `EndpointParser`,

`EndpointParser(name="completions").parse("/some_deployment/chat/completions") will parse successfully, with
`deployment_id = some_deployment/chat`

Reply:
The approach here, other than I use right now (check that chat completions is not valid in same statement, to filter chat completions from completions) is to create subclass

```python
class CompletionsEndpointParser(EndpointParser):
    ...
```
to exclude `chat/completions` case

Reply:
> This would not work with current parsers
The parsers aren't set in stone.

This condition looks so unnatural, because we didn't bother to fix the parsers when `/completions` case was introduced. Conceptually parsers should be disjoint.